File: java/Day2.java
import java.io.*;
import java.util.*;

public class Day2 {
    public static void main(String[] args) throws FileNotFoundException {
        long startTime = System.currentTimeMillis();
        String file = "input/day2.txt";
        Scanner in = new Scanner(new FileReader(file));
        List<String> input = new ArrayList<>();
        while (in.hasNextLine()) {
            input.add(in.nextLine());
        }
        System.out.println("Part I : " + isPossible(input));
        System.out.println("Part II: " + minPower(input));
        System.out.println("Execution time: " + (System.currentTimeMillis() - startTime) + "ms");
    }
    
    static Map<String, Integer> cubes;
    
    private static int isPossible(List<String> in) {
        cubes = new HashMap<>() {{
            put("red", 12);
            put("green", 13);
            put("blue", 14);
        }};
        int id = 1;
        int res = 0;
        for (String line : in) {
            if (checkLine(line)) {
                res += id;
            }
            id++;
        }
        return res;
    }
    
    private static boolean checkLine(String line) {
        String[] draws = line.substring(line.indexOf(":") + 1).trim().split("; ");
        for (String draw : draws) {
            for (String token : draw.split(", ")) {
                String key = token.split(" ")[1];
                int ct = Integer.parseInt(token.split(" ")[0]);
                if (ct > cubes.get(key)) {
                    return false;
                }
            }
        }
        return true;
    }
    
    private static int minPower(List<String> in) {
        int res = 0;
        for (String line : in) {
            Map<String, Integer> map = new HashMap<>() {{
                put("red", 0);
                put("green", 0);
                put("blue", 0);
            }};
            String[] draws = line.substring(line.indexOf(":") + 1).trim().split("; ");
            for (String draw : draws) {
                for (String token : draw.split(", ")) {
                    String key = token.split(" ")[1];
                    int ct = Integer.parseInt(token.split(" ")[0]);
                    map.put(key, Math.max(ct, map.get(key)));
                }
            }
            int sum = 1;
            for (int i : map.values()) {
                sum *= i;
            }
            res += sum;
        }
        return res;
    }
}

File: rust/src/bin/day2.rs
use std::{fs, collections::HashMap};

lazy_static::lazy_static! {static ref FILE:String = fs::read_to_string("../input/day2.txt").unwrap();}

fn part1(input: &str, cubes: &HashMap<String, i32>) -> i32 {
    let mut sum: i32 = 0;
    for line in input.lines() {
        let parts: Vec<&str> = line.splitn(2, ":").collect();
        let id: i32 = parts[0].trim().split_whitespace().last().unwrap().parse().unwrap();
        let tokens = parts[1].trim().split(";");
        let mut possible = true;
        for token in tokens {
            let draws = token.trim().split(",");
            for draw in draws {
                let mut iter = draw.split_whitespace();
                if let (Some(count_str), Some(label)) = (iter.next(), iter.next()) {
                    let count: i32 = count_str.parse().unwrap();
                    if count > cubes[label] {
                        possible = false;
                    }
                }
            }
            if !possible {
                break;
            }
        }
        if possible {
            sum += id;
        }
    }
    sum
}

fn part2(input: &str) -> i32 {
    let mut res = 0;
    for line in input.lines() {
        let mut map: HashMap<String, i32> = HashMap::new();
        map.insert("red".to_string(), 0);
        map.insert("green".to_string(), 0);
        map.insert("blue".to_string(), 0);

        if let Some(pos) = line.find(":") {
            let draws = line[pos + 1..].trim().split(";");
            for draw in draws {
                for token in draw.split(", ") {
                    let parts: Vec<&str> = token.split_whitespace().collect();
                    if parts.len() == 2 {
                        if let Ok(ct) = parts[0].parse::<i32>() {
                            let key = parts[1].to_string();
                            map.entry(key).and_modify(|e| *e = i32::max(*e, ct));
                        }
                    }
                }
            }
        }
        let sum: i32 = map.values().product();
        res += sum;
    }
    res
}

fn get_cubes() -> HashMap<String, i32> {
    let mut cubes = HashMap::new();
    cubes.insert("red".to_string(), 12);
    cubes.insert("green".to_string(), 13);
    cubes.insert("blue".to_string(), 14);
    cubes
}

fn main() {
    let mut cubes = get_cubes();
    println!("{}{}", "Expect 2879, Result: ", part1(&FILE, &cubes));
    println!("{}{}", "Expect 65122, Result: ", part2(&FILE));
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn day2_part1() {
        let cubes = get_cubes();
        assert_eq!(2879, part1(&FILE, &cubes))
    }

    #[test]
    fn day2_part2() {
        let cubes = get_cubes();
        assert_eq!(65122, part2(&FILE))
    }
}

